benches/bin/utils.rs

```diff
@@ -0,0 +1,109 @@
+extern crate rand;
+
+use std::sync::atomic::{ATOMIC_U64_INIT, AtomicU64, Ordering};
+use std::time::{Duration, Instant};
+use std::ops::Div;
+use std::iter::Sum;
+use tikv::coprocessor::codec::table::{encode_index_seek_key, encode_row_key};
+use tikv::util::codec::number::NumberEncoder;
+use test::black_box;
+
+use rand::Rng;
+
+#[inline]
+pub fn next_ts() -> u64 {
+    static CURRENT: AtomicU64 = ATOMIC_U64_INIT;
+    CURRENT.fetch_add(1, Ordering::SeqCst)
+}
+
+/// Generate `count` row keys that all with the specified `table_id`.
+///
+/// Row id will start with `start_id` and increment
+pub fn generate_row_keys(table_id: i64, start_id: i64, count: usize) -> Vec<Vec<u8>> {
+    let mut result = Vec::with_capacity(count);
+    for i in (start_id)..(start_id + count as i64) {
+        let mut handle = Vec::with_capacity(8);
+        handle.encode_i64(i as i64).unwrap();
+        let key = encode_row_key(table_id, &handle);
+        result.push(key);
+    }
+    result
+}
+
+
+/// Generate `count` unique index keys that all with the specified `table_id`, `index_id`,
+/// and `value_len`.
+pub fn generate_unique_index_keys(
+    table_id: i64,
+    index_id: i64,
+    value_len: usize,
+    count: usize,
+) -> Vec<Vec<u8>> {
+    if value_len < 8 {
+        panic!("Value len should be at least 8.");
+    }
+    let mut result = Vec::with_capacity(count);
+    for i in 0..count {
+        let mut value = Vec::with_capacity(value_len);
+        value.resize(value_len - 8, 0);
+        value.encode_u64(i as u64).unwrap();
+        result.push(encode_index_seek_key(table_id, index_id, &value));
+    }
+    result
+}
+
+pub fn shuffle<T>(data: &mut [T]) {
+    let mut rng = rand::thread_rng();
+    for i in 0..(data.len()) {
+        let j = rng.gen_range(i, data.len());
+        data.swap(i, j);
+    }
+}
+
+
+/// Convert duration to nanoseconds
+pub fn to_total_nanos(duration: &Duration) -> u64 {
+    duration.as_secs() * 1_000_000_000 + (duration.subsec_nanos() as u64)
+}
+
+/// Run `job` for `iterations` times, and return a Vec containing nanoseconds of each turn `job`
+/// returns.
+///
+/// `job` must return a `Duration` type. Please calculate time cost of what you want to bench
+/// manually ant return it.
+///
+/// Attention that too short job may cause great inaccuracy.
+pub fn record_time<F>(mut job: F, iterations: u32) -> Vec<u64>
```

Comment:
It's better to calculate time once after all jobs finished, not for each job once.